pkg/backend/httpstate/client/client.go

```diff
@@ -1361,3 +1369,79 @@ func (pc *Client) SubmitAIPrompt(ctx context.Context, requestBody interface{}) (
 	res, err := pc.do(ctx, request)
 	return res, err
 }
+
+type PublishPackageInput struct {
+	// Source is the source of the package. Typically this is 'pulumi' for packages published to the Pulumi Registry.
+	// Packages loaded from other registries (e.g. 'opentofu') will point to the origin of the package.
+	Source string
+	// Publisher is the organization that is publishing the package.
+	Publisher string
+	// Name is the URL-safe name of the package.
+	Name string
+	// Version is the semantic version of the package that should get published.
+	Version semver.Version
+	// Schema is a reader containing the JSON schema of the package.
+	Schema io.Reader
+	// Readme is a reader containing the markdown content of the package's README.
+	Readme io.Reader
+	// InstallDocs is a reader containing the markdown content of the package's installation documentation.
+	// This is optional, and if omitted, the package will not have installation documentation.
+	InstallDocs io.Reader
+}
+
+func (pc *Client) PublishPackage(ctx context.Context, input PublishPackageInput) error {
+	req := apitype.StartPackagePublishRequest{
+		Version: input.Version.String(),
+	}
+	var resp apitype.StartPackagePublishResponse
+	err := pc.restCall(ctx, "POST", publishPackagePath(input.Source, input.Publisher, input.Name), nil, req, &resp)
+	if err != nil {
+		return fmt.Errorf("publish package failed: %w", err)
+	}
+
+	uploadFile := func(url string, reader io.Reader, fileType string) error {
+		putReq, err := http.NewRequest(http.MethodPut, url, reader)
+		if err != nil {
+			return fmt.Errorf("failed to upload %s: %w", fileType, err)
+		}
+		for k, v := range resp.RequiredHeaders {
+			putReq.Header.Add(k, v)
+		}
```

Comment:
Why is this in the response to the initial post request?  Shouldn't the CLI know what headers it sets based on what it uploads?  It seems strange to me that the server tells us what headers we need to use.

Reply:
To use signed URLs with Azure Blob Storage, the header `x-ms-blob-type` must be set to `BlockBlob` by the client using the signed URL.
But the client doesn't know what kind of blob storage it's interacting with. The server does though.

We're doing the same for policy packs btw